include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,1073 @@
+ *
+ * std::vector<std::complex<double>> displacement = {...};
+ *
+ * data.write_dataset("displacement", displacement);
+ *
+ * // Write the simulation metadata
+ * data.set_attribute("active_cells", triangulation.n_active_cells());
+ * @endcode
+ *
+ * Read the simulation results with python:
+ * ~~~~~~~~~~~~~{.py}
+ * h5_file = h5py.File('simulation.hdf5','r+')
+ * data = h5_file['data']
+ * displacement = data['displacement'] # complex128 dtype
+ * active_cells = data.attrs['degrees_of_freedom'])
+ * ~~~~~~~~~~~~~
+ *
+ * # HDF5 and thread safety
+ * By default HDF5 is not thread-safe. The HDF5 library can be configured to be
+ * thread-safe, see [the HDF5
+ * documentation](https://support.hdfgroup.org/HDF5/faq/threadsafe.html). The
+ * thread-safe HDF5 version serializes the API but does not provide any level of
+ * concurrency. To achieve high parallel performance with HDF5, we advice to use
+ * HDF5 with MPI.
+ *
+ * @author Daniel Garcia-Sanchez, 2018
+ */
+// clang-format on
+namespace HDF5
+{
+  /**
+   * Base class for the HDF5 objects.
+   *
+   * @author Daniel Garcia-Sanchez, 2018
+   */
+  class HDF5Object
+  {
+  protected:
+    /**
+     * Constructor. @p string_name is the name of the HDF5 Object. If @p mpi is
+     * True then MPI I/O is used.
+     */
+    HDF5Object(const std::string name, const bool mpi);
+
+  public:
+    /**
+     * Reads an attribute. @p T can be `float`, `double`, `std::complex<float>`,
+     * `std::complex<double>`, `int`, `unsigned int`, `bool` or `std::string`.
+     * Note that the encoding of `std::string` is UTF8 in order to be compatible
+     * with python3.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <typename T>
+    T
+    get_attribute(const std::string &attr_name) const;
+
+    /**
+     * Writes an attribute. @p T can be `float`, `double`, `std::complex<float>`,
+     * `std::complex<double>`, `int`, `unsigned int`, `bool` or `std::string`.
+     * Note that the encoding of `std::string` is UTF8 in order to be compatible
+     * with python3.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <typename T>
+    void
+    set_attribute(const std::string &attr_name, const T value);
+
+
+    const std::string name;
```

Comment:
This is a `public` variable. Is this intended? Our style would ordinarily be to make this variable `private` or `protected`, and if necessary provide a `get_name()` function.

In any case, you probably want to document it.